feat(lsp)!: change handler signature

mjlbach opened the pull request:
Previously, the handler signature was:

  function(err, method, params, client_id, bufnr, config)

In order to better support external plugins that wish to extend the
protocol, there is other information which would be advantageous to
forward to the client, such as the original params of the request that
generated the callback.

In order to do this, we would need to break symmetry of the handlers, to
add an additional "params" as the 7th argument.

Instead, this PR changes the signature of the handlers to:

  function(err, result, context, config)

where context includes params, client_id, and bufnr. This also leaves
flexibility for future use-cases.

Todos:
- [x] fix tests
- [x] make sure I didn't break anything
- [x] update docs
- [x] warn plugin authors
- [x] Decide what to do about compatibility question

mjlbach committed:
feat(lsp)!: change handler signature

Previously, the handler signature was:

  function(err, method, params, client_id, bufnr, config)

In order to better support external plugins that wish to extend the
protocol, there is other information which would be advantageous to
forward to the client, such as the original params of the request that
generated the callback.

In order to do this, we would need to break symmetry of the handlers, to
add an additional "params" as the 7th argument.

Instead, this PR changes the signature of the handlers to:

  function(err, result, ctx, config)

where ctx (the context) includes params, client_id, and bufnr. This also leaves
flexibility for future use-cases.

BREAKING_CHANGE: changes the signature of the built-in client handlers, requiring
updating handler calls

jose-elias-alvarez commented:
Looks great! Do you think it would make sense to handle #15174 here, too, since the PR modifies the interaction between `buf_request` and `buf_request_all`? As far as I can tell, the implementation here will run into the same `nil` index issue, since `context.client_id` will be `nil` if no client supports a given method. (My [PR](https://github.com/neovim/neovim/pull/15193) to fix that issue seems to have stalled and is incompatible with the changes here, anyways.)

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1316,8 +1316,8 @@ function lsp.buf_request_all(bufnr, method, params, callback)
 
-  local function _sync_handler(err, _, result, client_id)
-    request_results[client_id] = { error = err, result = result }
+  local function _sync_handler(err, result, context)
```
naming nit: any objection to using `ctx` for this arg everywhere? it's going to be all over the place I guess, and would rather establish that convention now. `ctx` is a common convention for "context" and has the mild benefit of disambiguating unrelated uses of the word "context" e.g. with `grep ctx`.

mjlbach replied:
I'm fine with it, wdyt @neovim/lsp (@tjdevries)?

mfussenegger replied:
Sounds good to me :+1:

mjlbach commented:
> Looks great! Do you think it would make sense to handle #15174 here, too, since the PR modifies the interaction between `buf_request` and `buf_request_all`? As far as I can tell, the implementation here will run into the same `nil` index issue, since `context.client_id` will be `nil` if no client supports a given method. (My [PR](https://github.com/neovim/neovim/pull/15193) to fix that issue seems to have stalled and is incompatible with the changes here, anyways.)

I think that is out of scope of this PR, but I will rebase and review your PR after this.

mjlbach committed:
docs: regenerate

mjlbach commented:
A couple notes, I changed `result` to `params` in a few places for "consistency" but maybe I should not have done this, as `params` is codified in the spec. LMK what you think.

mjlbach commented:
@ray-x @jose-elias-alvarez @hrsh7th @glepnir @akinsho @simrat39 this is a warning that (although we will not be merging this yet) it will break your plugins that rely on overriding handlers (also lmk if there were any obvious ones I missed)

ms-jpq commented:
this breaks existing `buf_request` handlers right?

mjlbach commented:
@ms-jpq I knew there was someone I forgot :) 

The function signature for buf_request itself is the same, but the handler argument follows the new pattern.

ms-jpq commented:
thanks :)

so `config` is always a table and never nil right? then we can just check how many elements are in the function argument?, and do it that way.

no version detection required

mjlbach commented:
@ms-jpq config can be nil, but if nil is set to an empty table within the handler.  I'd recommend using the wrapper I added in vim.lsp.compat and `vim.fn.has('nvim-0.5.1')` to gate which handler paradigm you use, we'll push this change to 0.5.1 so it shouldn't be *too* disruptive (hopefully...)

ms-jpq commented:
haha yeah, but it would break all the things for people who havn't updated the plugins though.

I think Ill just use `bufnr` on arg position `5` as a gate, then you are free to do whatever with the version numbers 👍

ms-jpq commented:
anyways, thank you for letting us know early 💯

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -244,15 +249,18 @@ For |lsp-notification|, each |lsp-handler| has this signature: >
-                            The ID of the |vim.lsp.client|
-            {bufnr}     (nil)
-                            `nil`, as the server doesn't have an associated buffer.
+            {context}   (table)
```
ctx? also the function signature on line 246 wasn't updated.

mjlbach replied:
Fixed

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -672,13 +672,13 @@ function lsp.start_client(config)
-  ---@param params (table) The parameters for that method.
-  function dispatch.notification(method, params)
-    local _ = log.debug() and log.debug('notification', method, params)
+  ---@param result (table) The parameters for that method.
```
is this docstring correct? result=parameters ?

mjlbach replied:
Changed this signature back

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -687,12 +687,12 @@ function lsp.start_client(config)
   ---@param params (table) The parameters for that method
-  function dispatch.server_request(method, params)
-    local _ = log.debug() and log.debug('server_request', method, params)
+  function dispatch.server_request(method, result)
```
docstring outdated

mjlbach replied:
I changed it back, to params, I think result make sense in the signature of our handlers, but we should stick with params in the dispatch function table, as that maps directly onto the spec (my bad for changing it originally in the PR, see my above comment earlier in the thread.

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1316,8 +1316,8 @@ function lsp.buf_request_all(bufnr, method, params, callback)
 
-  local function _sync_handler(err, _, result, client_id)
-    request_results[client_id] = { error = err, result = result }
+  local function _sync_handler(err, result, context)
```
```suggestion
  local function _sync_handler(err, result, ctx)
```

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1316,8 +1316,8 @@ function lsp.buf_request_all(bufnr, method, params, callback)
-  local function _sync_handler(err, _, result, client_id)
-    request_results[client_id] = { error = err, result = result }
+  local function _sync_handler(err, result, context)
+    request_results[context.client_id] = { error = err, result = result }
```
```suggestion
    request_results[ctx.client_id] = { error = err, result = result }
```

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1498,8 +1498,8 @@ end
 function lsp.with(handler, override_config)
-  return function(err, method, params, client_id, bufnr, config)
-    return handler(err, method, params, client_id, bufnr, vim.tbl_deep_extend("force", config or {}, override_config))
+  return function(err, result, context, config)
+    return handler(err, result, context, vim.tbl_deep_extend("force", config or {}, override_config))
```
```suggestion
  return function(err, result, ctx, config)
    return handler(err, result, ctx, vim.tbl_deep_extend("force", config or {}, override_config))
```

justinmk commented on runtime/lua/vim/lsp/compat.lua:
```diff
@@ -0,0 +1,12 @@
+
+M.wrap_05_handler = function(handler)
+  return function(err, result, context, config)
+    return handler(err, context.method, result, context.client_id, context.bufnr, config)
```
this looks like something that can be copy-pasted into any project. So we can document it, but we don't need to introduce this compat module into the API and then later deal with removing it (thus duplicating  the migration issue we're already trying to deal with now)

mfussenegger replied:
I was just thinking how to adapt the handlers in my plugins so that they work with all three: 0.5, master prior this change and master with this change.

And that would require some kind of feature or version check.
With the current PR state I'd probably create something like this:


```lua
function M.mk_handler(fn)
  if vim.lsp.compat then
    return vim.lsp.compat.wrap_05_handler(fn)
  end
  return fn
end
```

Implementing the `wrap_05_handler` isn't the issue, but some kind of marker that can be checked at runtime to see which signature format is used would be good. `nvim-0.6` won't be granular enough, as it would also return true for versions after 0.5, but before this change.

justinmk replied:
> `nvim-0.6` won't be granular enough, as it would also return true for versions after 0.5, but before this change.

? `:echo has('nvim-0.5.1')` works

mfussenegger replied:
`vim.fn.has('nvim-0.5.1')` would also return true for the current master (without this change, using the old signature) 

I guess it is acceptable that master/plugin combinations are broken for a bit, but if there were some check for patch-level or signature-format level plugins could make the transition seamless. (At least if they adapt the code before this PR is merged)

mfussenegger replied:
Something like the following might work to keep compatiblity with all 3 variants:

```
function M.mk_handler(fn)
  return function(...)
    local is_new = type(select(4, ...)) == 'table'
    if is_new then
      fn(...)
    else
      local err = select(1, ...)
      local method = select(2, ...)
      local result = select(3, ...)
      local client_id = select(4, ...)
      local bufnr = select(5, ...)
      local config = select(6, ...)
      fn(err, result, { method = method, client_id = client_id, bufnr = bufnr }, config)
    end
  end
end
```

mfussenegger commented on runtime/lua/vim/lsp/handlers.lua:
```diff
@@ -451,9 +457,12 @@ end
+    local method = ctx.method
+    local bufnr = ctx.bufnr
     local _ = log.debug() and log.debug('default_handler', method, {
-      params = params, client_id = client_id, err = err, bufnr = bufnr, config = config
+      result = result, client_id = client_id, err = err, bufnr = bufnr, config = config
     })
```
Might be worth logging the context with vim.inspect, to ensure this automatically logs everything if the context gets extended.

mjlbach replied:
vim.inspect has a runtime cost though right? I guess it's fine since this is only done if the user is using debug

mfussenegger replied:
I'd think the overhead is acceptable given that it is logged on debug level and that is not enabled by default.

mjlbach commented:
I think I have addressed all of the comments.

The main remaining to-do is to figure out the compatibility question:

1. Do we include a compat module (originally suggested by TJ) or just document here/in breaking changes issue how to work around this in plugins?
2. Do we do a runtime compatibility check somehow for testing if users are using old argument style (possible runtime cost)
3. How long do we wait after this warning to plugin authors? Certain plugins (lspsaga) are not likely to be updated.


I think we can cut 0.5.1 after this (from an LSP perspective at least), however, we either need to fix, or not backport the URI changes that caused https://github.com/neovim/neovim/issues/15261

justinmk commented:
> 1. Do we include a compat module (originally suggested by TJ) or just document here/in breaking changes issue how to work around this in plugins?

Even with a compat module, we need to document steps that extension authors must perform, right? So the only difference is that without the compat module, we have one less thing to deprecate.

> 2. Do we do a runtime compatibility check somehow for testing if users are using old argument style (possible runtime cost)

It's a temporary cost and hopefully minimal code, so why not? We can remove the runtime check in 0.5.2+.

> 3. How long do we wait after this warning to plugin authors to merge this PR? Certain plugins (lspsaga) are not likely to be updated. I propose 1 week from today.

Merging to master doesn't really need to wait. It's more a question of how long we keep the runtime check, imo.

mjlbach commented:
There's unfortunately not a good way to do the runtime check. I thought I could do a check using `debug.getinfo(handler).nparams` on handlers, but unfortunately under the old schema handlers with 4 args are allowed, so there is no way to disambiguate old vs new (we can't check on `method` since we actually need to inject the check **into** the user defined handler, we can't wrap it with the check)

justinmk commented:
> There's unfortunately not a good way to do the runtime check. I thought I could do a check using `debug.getinfo(handler).nparams` on handlers,

Keep it simple. Can we just check the form of one of the args? 

The old signature was 

    function(err, method, result, client_id, bufnr, config)

the new signature is

    function(err, result, ctx, config)

can't we check if `ctx.client_id` is missing or some other field that we always expect on `ctx` ?

mjlbach commented:
No, because the issue is that we would have to inject that check *into* the handler that a user is registering, the issue is we only control which arguments are passed into the handler :/

ray-x commented:
I believe `textDocument/codeLens` handler update is missing

mjlbach commented:
@ray-x should be fixed now, thanks!

gebner commented:
> (also lmk if there were any obvious plugin authors I missed)

We have quite a few request and notification handlers in the lean.nvim plugin. cc @rish987 @Julian 

If I understand this correctly, we should be using a `mk_handler` function like in https://github.com/mfussenegger/nvim-lsp-compl/commit/29a81f3747373f42d002775b1f79dcd19fada2e4 to support both 0.5.0 and master.  I believe plenary.nvim needs to be updated as well (i.e., the async wrapper for buf_request).

mjlbach commented:
@hrsh7th @ray-x  gentle ping on this. Can people generally let me know the state of their plugins? I'd like to merge this in the next several days

jose-elias-alvarez commented:
null-ls and friends are good to go.

hrsh7th commented:
@mjlbach Thank you for your mention.

I'm thinking to implement the following compact code.

```lua
request(function(arg1, arg2, arg3, arg4, arg5 ,arg6)
  if type(arg3) == 'table' and type(arg3.client_id) == 'number' then
    -- NEW: err, result, context, config
    local err, result, context, config = arg1, arg2, arg3, arg4
  else
    -- OLD: err, method, params, client_id, bufnr, config
    local err, method, params, client_id, bufnr, config = arg1, arg2, arg3, arg4, arg5, arg6
  end
end)
```

Is this correct?

hrsh7th commented:
I've added the compatible codes https://github.com/hrsh7th/cmp-nvim-lsp/commit/806c05dd4278fae551794bf8b0c5f692c63b620c :)

mjlbach commented:
Ok, great. I am planning to merge this at 1600 (4 pm) utc on September 5th. Please let me know if there are any imminent blockers!

mjlbach commented:
1 hour grace period from 4 pm :) lspsaga is not updated, and plenary is not yet either, but should be good otherwise

justinmk commented:
- Did we add https://github.com/neovim/neovim/pull/15504#discussion_r698074955 to the `:help lsp` so that there is a helptag we (and users) can reference whenever the question is raised?
- Also consider pasting that compat pattern into https://github.com/neovim/neovim/issues/14090#issuecomment-913198455

mjlbach commented:
I added a note in the breaking changes issue, but I can file a PR adding some documentation.

justinmk commented:
@mjlbach can you send this also to the 0.5-release branch
